### PR TITLE
Refactor FilePrefetchBuffer code

### PR DESCRIPTION
**Pull request by akankshamahajan15**
Summary - Refactor FilePrefetchBuffer code 
- Implementation:
FilePrefetchBuffer maintains a deque of free buffers (free_bufs_) of size num_buffers_ and buffers (bufs_) which contains the prefetched data. Whenever a buffer is consumed or is outdated (w.r.t. to requested offset), that buffer is cleared and returned to free_bufs_.

 If a buffer is available in free_bufs_, it's moved to bufs_ and is sent for prefetching. num_buffers_ defines how many buffers are maintained that contains prefetched data.
If num_buffers_ == 1, it's a sequential read flow. Read API will be called on that one buffer whenever the data is requested and is not in the buffer.
If num_buffers_ > 1, then the data is prefetched asynchronosuly in the buffers whenever the data is consumed from the buffers and that buffer is freed.
If num_buffers > 1, then requested data can be overlapping between 2 buffers. To return the continuous buffer overlap_bufs_ is used. The requested data is copied from 2 buffers to the overlap_bufs_ and overlap_bufs_ is returned to
the caller.

- Merged Sync and Async code flow into one in FilePrefetchBuffer.

Test Plan - 
- Crash test passed
- Unit tests
- Pending - Benchmarks

### CODE REVIEW

File: file/file_prefetch_buffer.h [hunk 1]
```diff
@@ -31,7 +32,37 @@ namespace ROCKSDB_NAMESPACE {
...
+
+  // If true, Readahead is enabled implicitly by rocksdb
+  // after doing sequential scans for num_file_reads_for_auto_readahead.
+  bool implicit_auto_readahead = false;
+
+  uint64_t num_file_reads = 0;
+  uint64_t num_file_reads_for_auto_readahead = 0;
+
+  // Number of buffers to maintain that contains prefetched data. If num_buffers
+  // > 2 then buffers will be filled asynchronously whenever they get emptied.
```
Comment (akankshamahajan15) [inline file/file_prefetch_buffer.h:56]: TODO - Replace 2 with 1

File: file/file_prefetch_buffer.h [hunk 2]
```diff
@@ -60,129 +99,146 @@ enum class FilePrefetchBufferUsage {
...
     assert((num_file_reads_ >= num_file_reads_for_auto_readahead_ + 1) ||
            (num_file_reads_ == 0));
-    // If ReadOptions.async_io is enabled, data is asynchronously filled in
-    // second buffer while curr_ is being consumed. If data is overlapping in
-    // two buffers, data is copied to third buffer to return continuous buffer.
-    bufs_.resize(3);
-    for (uint32_t i = 0; i < 2; i++) {
-      bufs_[i].pos_ = i;
+
+    // If num_buffers_ > 2, data is asynchronously filled in the
```
Comment (akankshamahajan15) [inline file/file_prefetch_buffer.h:168]: Replace 2 with 1

File: table/block_based/block_prefetcher.cc
Comment (ajkr) [inline table/block_based/block_prefetcher.cc:1]: Idea for another refactoring/behavior change: move functionality out of `BlockPrefetcher` and eliminate it. `BlockPrefetcher` appears to perform some mix of: (1) monitor file reads for a `BlockBasedTableIterator`, (2) create internal readahead buffers, and (3) issue filesystem readahead. Each of these functions feels a little bit odd to happen in this layer (not to mention in a class called `BlockPrefetcher`).

Take (1) for example. We claim to count the number of file reads by incrementing a counter called `num_file_reads_` in a function called `PrefetchIfNeeded()`. The function `PrefetchIfNeeded()` is called when initializing any data block, whether it requires a file read or not (it may be in block cache).

Or (2), for example. The function `PrefetchIfNeeded()` delegates responsibility for actually prefetching to another class (`FilePrefetchBuffer`). So prefetching may be needed but it doesn't happen here.

(3) is confusing just because internal prefetching happens in such a distant call stack (`BlockBasedTableIterator::{Async,}InitDataBlock()` -> ... (BlockBasedTable, BlockFetcher, FilePrefetchBuffer) ... -> `FilePrefetchBuffer::Prefetch()`) compared to external prefetching (`BlockBasedTableIterator::{Async,}InitDataBlock()` -> `BlockPrefetcher::PrefetchIfNeeded()`). I don't know why it needs to be so far apart.

**Review by ajkr — CHANGES_REQUESTED** [review]
Glad to see this refactoring! I made it through everything except file_prefetch_buffer.cc so far

File: table/block_based/block_prefetcher.cc
Comment (akankshamahajan15) [reply table/block_based/block_prefetcher.cc:1]: Yes!!! I'm planning to remove BlockPrefetcher from the equation. However I didn't want to include it in this diff. I'm planning to address it in a separate PR. This PR is more about refactoring FilePrefetchBuffer as I didn't want to overwhelm this PR with all the refactoring. I've added BlockPrefetcher refactoring in our roadmap as well.

**Comment by akankshamahajan15** [comment]
Thanks a lot for the review and comments @ajkr. I'll go through them and 'll address them.

File: file/file_prefetch_buffer.h [hunk 3]
```diff
@@ -487,10 +488,104 @@ class FilePrefetchBuffer {
...
+      }
+    }
+    bufs_ = tmp_buf;
+  }
+
+  // *** END APIs related to allocating and freeing buffers ***
+
+  std::deque<BufferInfo*> bufs_;
+  std::deque<BufferInfo*> free_bufs_;
+  std::deque<BufferInfo*> overlap_bufs_;
```
[the remark this answers is not in the document]
Comment (akankshamahajan15) [reply file/file_prefetch_buffer.h:588]: Deque works better because it's easy to remove from the front and back. Let me see If I can make it queue since now implementation has been finalized. Let me know If you have any concerns with Queue.

File: file/file_prefetch_buffer.h [hunk 4]
```diff
@@ -77,138 +104,158 @@ enum class FilePrefetchBufferUsage {
   kUnknown,
 };
 
+// Implementation:
+// FilePrefetchBuffer maintains a dequeu of free buffers (free_bufs_) of size
+// num_buffers_ and bufs_ which contains the prefetched data. Whenever a buffer
```
[the remark this answers is not in the document]
Comment (akankshamahajan15) [reply file/file_prefetch_buffer.h:109]: Yes it's num_buffers_ == bufs_.size() + free_bufs_.size(). I'll update the comments to make it more clear.

File: file/file_prefetch_buffer.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajkr) [reply file/file_prefetch_buffer.h:588]: Oh sorry I meant to comment only on `overlap_bufs_`. GitHub shows the context so it might have looked like I was commenting on all three uses of `deque`. The reason I thought `overlap_bufs_` may not need to be a `deque` is because it appears to always have zero or one entry, in which case it wouldn't need to be a collection data type.

[the remark this answers is not in the document]
Comment (akankshamahajan15) [reply file/file_prefetch_buffer.h:588]: Ohh okk! Yeah that makes sense. I can change that. Thanks

File: file/file_prefetch_buffer.cc [hunk 5]
```diff
@@ -399,289 +376,285 @@ void FilePrefetchBuffer::ReadAheadSizeTuning(
...
+    // only if requested data lies within next buffer.
+    size_t second_size = next_buf->async_read_in_progress_
+                             ? next_buf->async_req_len_
+                             : next_buf->CurrentSize();
+    uint64_t start_offset = next_buf->initial_end_offset_;
     // Second buffer might be out of bound if first buffer already prefetched
     // that data.
-    if (tmp_offset + tmp_length <= bufs_[second].offset_ + second_size) {
+
+    if (tmp_offset + tmp_length <= next_buf->offset_ + second_size) {
```
Comment (ajkr) [inline file/file_prefetch_buffer.cc:453]: What does it mean for `tmp_offset + tmp_length > next_buf->offset_ + second_size`? The read is not done but we're also not going to read more?

I also couldn't understand the comment: "Second buffer might be out of bound ...". What bounds are we referring to?

File: file/file_prefetch_buffer.cc [hunk 6]
```diff
@@ -248,90 +241,74 @@ void FilePrefetchBuffer::AbortAllIOs() {
...
-               bufs_[curr_].offset_ + bufs_[curr_].buffer_.CurrentSize())) {
-            bufs_[second].ClearBuffer();
+  BufferInfo* buf = GetFirstBuffer();
+
+  // In case buffers do not align, reset next buffer if requested data needs
+  // to be read in that buffer.
+  if (NumBuffersAllocated() > 1) {
+    BufferInfo* next_buf = bufs_[1];
+    if (!next_buf->async_read_in_progress_ && !buf->async_read_in_progress_ &&
+        buf->DoesBufferContainData()) {
```
Comment (ajkr) [inline file/file_prefetch_buffer.cc:278]: Maybe add `assert(bufs_.size() <= 2);` in the places it's assumed. I think this is one such place since it doesn't look for misalignment between consecutive `bufs_` after the second one

**Review by ajkr — COMMENTED** [review]
Please feel free to leave [existing issue] comments for another time; I just recorded them here since reviewing this PR made them visible.

**Comment by akankshamahajan15** [comment]
NOTE : Some of the stress tests are failing for async_io after refactoring. Once that's fixed (most likely tomorrow), I'll address the comments and update the diff on Wednesday.

File: file/file_prefetch_buffer.cc (hunk 5 above)
Comment (akankshamahajan15) [reply file/file_prefetch_buffer.cc:453]: Oh let me update the comment. I forgot to remove this comment.

**Review by ajkr — COMMENTED** [review]
Finished a complete pass.

File: file/file_prefetch_buffer.cc [hunk 7]
```diff
@@ -120,17 +115,15 @@ Status FilePrefetchBuffer::ReadAsync(const IOOptions& opts,
...
-  req.scratch = bufs_[index].buffer_.BufferStart();
-  bufs_[index].async_req_len_ = req.len;
+  req.scratch = buf->buffer_.BufferStart();
+  buf->async_req_len_ = req.len;
 
-  Status s =
-      reader->ReadAsync(req, opts, fp, &(bufs_[index].pos_),
-                        &(bufs_[index].io_handle_), &(bufs_[index].del_fn_),
-                        /*aligned_buf=*/nullptr);
+  Status s = reader->ReadAsync(req, opts, fp, buf->pos_, &(buf->io_handle_),
```
[the remark this answers is not in the document]
Comment (akankshamahajan15) [reply file/file_prefetch_buffer.cc:121]: That should definitely work. I can give it a try.

File: file/file_prefetch_buffer.cc (hunk 5 above)
Comment (akankshamahajan15) [reply file/file_prefetch_buffer.cc:464]: If requested bytes - tmp_offset + tmp_length are in next buffer, freed buffer can go for further prefetching.
If requested bytes are not in next buffer, next buffer has to go for sync call to get remaining requested bytes. In that it shouldn't go for async  prefetching as async prefetching calculates offset based on previous buffer end offset and previous buffer has to go for sync prefetching.

File: file/file_prefetch_buffer.h [hunk 8]
```diff
@@ -31,6 +32,31 @@ namespace ROCKSDB_NAMESPACE {
...
+  // on every IO until max_readahead_size is hit. Typically this is set as a
+  // multiple of initial_readahead_size. initial_readahead_size should be
+  // greater than equal to initial_readahead_size.
+  size_t max_readahead_size = 0;
+
+  // If true, Readahead is enabled implicitly by rocksdb
+  // after doing sequential scans for num_file_reads_for_auto_readahead.
+  bool implicit_auto_readahead = false;
+
+  uint64_t num_file_reads = 0;
```
[the remark this answers is not in the document]
Comment (akankshamahajan15) [reply file/file_prefetch_buffer.h:52]: Can't remove the num_file_reads_ because it's initialized and passed down from BlockPrefetcher and based on it's value we decide if we want to prefetch or not in FilePrefetchBuffer. It's value right now is either 0 or >= 2. However I will add a todo note to remove it when refactoring BlockPrefetcher. In that case we don't need to pass it down but we do need that variable.

File: file/file_prefetch_buffer.cc (hunk 6 above)
Comment (akankshamahajan15) [reply file/file_prefetch_buffer.cc:284]: I have added the check in all those places `(NumBuffersAllocated > 1)` or to return if NumBuffersAllocated() == 0 or 1.

Comment (akankshamahajan15) [reply file/file_prefetch_buffer.cc:284]: Here I added that condition at Line 281

File: file/file_prefetch_buffer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (ajkr) [reply file/file_prefetch_buffer.h:52]: Works for me

File: file/file_prefetch_buffer.cc [hunk 9]
```diff
@@ -399,289 +376,285 @@ void FilePrefetchBuffer::ReadAheadSizeTuning(
...
+
   Status s;
   size_t alignment = reader->file()->GetRequiredBufferAlignment();
-  uint32_t second;
+
+  BufferInfo* buf = GetFirstBuffer();
 
   // Check if the first buffer has the required offset and the async read is
   // still in progress. This should only happen if a prefetch was initiated
   // by Seek, but the next access is at another offset.
```
[the remark this answers is not in the document]
Comment (akankshamahajan15) [reply file/file_prefetch_buffer.cc:420]: Yes. So the data is overlapping between 2 buffers. Offset lies in first buffer so first buffer has the data but async is in progress for that buffer. So we call Poll here to get the data first and then copy the data to overlap_buf_.

We assumption was Poll is supposed to happen in `PrefetchInternal`. 
But for seek operation, we call `PrefetchAsync`  to submit async requests for all L0 files and L1-Ln levels. And then it calls again (second call) FilePrefetchBuffer to actually poll and read the data. It can be possible that second call wasn't made as data block was found in cache. So the buffer is never polled. 

This case handle that condition.
In short -  If buffer is needed but async is in progress, poll the data.

[the remark this answers is not in the document]
Comment (ajkr) [reply file/file_prefetch_buffer.cc:420]: > It can be possible that second call wasn't made as data block was found in cache.

The second call when there's a cache miss is like this, right? `SeekSecondPass()` -> ... -> `TryReadFromCache()` -> ... -> `PrefetchInternal()` 

I am seeing the code here poll even in that case, which is fine but seems to contradict the "next access is at another offset" part of the comment

File: file/file_prefetch_buffer.cc (hunk 5 above)
Comment (ajkr) [reply file/file_prefetch_buffer.cc:464]: It makes sense. I might think about how to write a loop that incorporates data from three or more async buffers, but haven't thought of anything to propose here yet.

File: file/file_prefetch_buffer.cc [hunk 10]
```diff
@@ -399,289 +382,291 @@ void FilePrefetchBuffer::ReadAheadSizeTuning(
...
+
+      // Length == 0: All the requested data has been copied to overlap buffer
+      // and it has already gone for async prefetching. It can return without
+      // doing anything further. Length > 0: More data needs to be consumed so
+      // it will continue async and sync prefetching and copy the remaining data
+      // to overlap buffer in the end.
+      if (length == 0) {
+        UpdateStats(/*found_in_buffer=*/true, length);
+        return s;
+      }
```
Comment (ajkr) [inline file/file_prefetch_buffer.cc:577]: I wonder if this belongs in `HandleOverlappingData()`. It's hard to say where that function should end and this one picks up. If we say logic specific to overlapping buffers belongs there, then we should probably move this.

**Review by ajkr — APPROVED** [review]
LGTM! I don't have any concerns but hope we will continue simplifying it

File: file/file_prefetch_buffer.cc (hunk 5 above)
Comment (ajkr) [reply file/file_prefetch_buffer.cc:464]: > think about how to write a loop that incorporates data from three or more async buffers

It's also useful to consolidate the handling for issuing more async readahead after freeing an outdated buffer. Right now I think the flow goes (1) free an outdated buffer (`ClearOutdatedData()`), (2) poll the front buffer, and then (3) allocate a new buffer and issue async readahead on it (`PrefetchRemBuffers()`). But if a buffer is consumed by `HandleOverlappingData()`, the new async readahead is issued before polling the next front buffer. The latter strategy has a little bit better I/O parallelism and I guess ideally there'd only be one strategy for both cases.
